packages/compass-data-modeling/src/store/relationships.ts

```diff
@@ -0,0 +1,215 @@
+import type { DataService } from '@mongodb-js/compass-connections/provider';
+import type { Document } from 'bson';
+import { isEqual } from 'lodash';
+import type { MongoDBJSONSchema } from 'mongodb-schema';
+import type { Relationship } from '../services/data-model-storage';
+
+/**
+ * A very simplistic depth-first traversing function that only handles a subset
+ * of real JSON schema keywords that is applicable to our MongoDB JSON schema
+ * format.
+ *
+ * Types are unwrapped: every bson type is treated as its own item to visit.
+ *
+ * Array items will have the same path as the array itself, mimicking how the
+ * paths would look like in mongodb query.
+ *
+ * @internal exported only for testing purposes
+ */
+export function traverseMongoDBJSONSchema(
+  schema: MongoDBJSONSchema,
+  visitor: (
+    schema: MongoDBJSONSchema,
+    path: string[],
+    isArrayItem: boolean
+  ) => void,
```

Comment:
I don't know if it's just me, but my first thought reading this function is "this callback could have been a generator". Obviously not really all that different in the end, but mentioning it in case you also feel that that would be more idiomatic